torchmultimodal/modules/layers/attention.py

```diff
@@ -0,0 +1,143 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+import itertools
+from typing import Optional
+
+import numpy as np
+
+import torch
+import torch.nn as nn
+
+from torchmultimodal.utils.common import tensor_slice
+
+
+class BroadcastPositionEmbedding(nn.Module):
+    r"""Spatiotemporal broadcasted positional embeddings.
+
+    Each embedding vector of the ``i``-th dim is repeated by ``N`` times, where
+    :math:`N = \prod_{j>i}\text{dim}[j]`.
+
+    Args:
+        shape (torch.Size): shape of raw data before batching and embedding
+        embedding_dim (int): the size of each embedding vector
+        dim (Optional[int]): dimension of embedding w.r.t. the other dimensions of batched data, e.g., ``-1``
+            corresponds to embedding being the last dimension. Default is ``-1``.
+
+    Raises:
+        ValueError: if ``dim`` is neither ``-1`` nor ``1``
+        ValueError: if ``embedding_dim`` is not an integer multiple of ``len(shape)``
+
+    Inputs:
+        x (Optional[torch.Tensor]): flattened input data, e.g., ``(batch, time * height * width, embedding_dim)``
+            if ``dim=-1``.
+        decode_step (Optional[int]): position of the data that requires decoding.
+    """
+
+    def __init__(
+        self, shape: torch.Size, embedding_dim: int, dim: Optional[int] = -1
+    ) -> None:
+        super().__init__()
+        if dim not in [-1, 1]:
+            raise ValueError(
+                f"Only first or last for the embedding dim is supported but got {dim}"
+            )
+        if embedding_dim % len(shape) != 0:
+            raise ValueError(
+                f"Embedding dim {embedding_dim} modulo len(shape) {len(shape)} is not zero"
+            )
+
+        self.shape = shape
+        self.n_dim = n_dim = len(shape)
+        self.embedding_dim = embedding_dim
+        self.dim = dim
+
+        self.embedding = nn.ParameterDict(
+            {
+                f"d_{i}": nn.Parameter(
+                    torch.randn(shape[i], embedding_dim // n_dim) * 0.01
+                    if dim == -1
+                    else torch.randn(embedding_dim // n_dim, shape[i]) * 0.01
+                )
+                for i in range(n_dim)
+            }
+        )
```

Comment:
nit: I think this way of using the comprehension makes this less readable. Would prefer to have the if/else on the outside